model/sighandler/sighandler_timestamp.go

```diff
@@ -0,0 +1,202 @@
+func (a *docTimeStamp) NewDigest(sig *model.PdfSignature) (model.Hasher, error) {
+	return bytes.NewBuffer(nil), nil
+}
+
+type timestampInfo struct {
+	Version        int
+	Policy         asn1.RawValue
+	MessageImprint struct {
+		HashAlgorithm pkix.AlgorithmIdentifier
+		HashedMessage []byte
+	}
+	SerialNumber    asn1.RawValue
+	GeneralizedTime time.Time
+}
+
+func getHashForOID(oid asn1.ObjectIdentifier) (crypto.Hash, error) {
+	switch {
+	case oid.Equal(pkcs7.OIDDigestAlgorithmSHA1), oid.Equal(pkcs7.OIDDigestAlgorithmECDSASHA1),
+		oid.Equal(pkcs7.OIDDigestAlgorithmDSA), oid.Equal(pkcs7.OIDDigestAlgorithmDSASHA1),
+		oid.Equal(pkcs7.OIDEncryptionAlgorithmRSA):
+		return crypto.SHA1, nil
+	case oid.Equal(pkcs7.OIDDigestAlgorithmSHA256), oid.Equal(pkcs7.OIDDigestAlgorithmECDSASHA256):
+		return crypto.SHA256, nil
+	case oid.Equal(pkcs7.OIDDigestAlgorithmSHA384), oid.Equal(pkcs7.OIDDigestAlgorithmECDSASHA384):
+		return crypto.SHA384, nil
+	case oid.Equal(pkcs7.OIDDigestAlgorithmSHA512), oid.Equal(pkcs7.OIDDigestAlgorithmECDSASHA512):
+		return crypto.SHA512, nil
+	}
+	return crypto.Hash(0), pkcs7.ErrUnsupportedAlgorithm
+}
+
+// Validate validates PdfSignature.
+func (a *docTimeStamp) Validate(sig *model.PdfSignature, digest model.Hasher) (model.SignatureValidationResult, error) {
+	signed := sig.Contents.Bytes()
+	p7, err := pkcs7.Parse(signed)
+	if err != nil {
+		return model.SignatureValidationResult{}, err
+	}
+
+	if err = p7.Verify(); err != nil {
+		return model.SignatureValidationResult{}, err
+	}
+
+	var tsInfo timestampInfo
+
+	_, err = asn1.Unmarshal(p7.Content, &tsInfo)
+	if err != nil {
+		return model.SignatureValidationResult{}, err
+	}
+
+	hAlg, err := getHashForOID(tsInfo.MessageImprint.HashAlgorithm.Algorithm)
+	if err != nil {
+		return model.SignatureValidationResult{}, err
+	}
+	h := hAlg.New()
+	buffer := digest.(*bytes.Buffer)
+
+	h.Write(buffer.Bytes())
+	sm := h.Sum(nil)
+	res := model.SignatureValidationResult{
+		IsSigned:        true,
+		IsVerified:      bytes.Equal(sm, tsInfo.MessageImprint.HashedMessage),
+		GeneralizedTime: tsInfo.GeneralizedTime,
+	}
+	return res, nil
+}
+
+// Sign sets the Contents fields for the PdfSignature.
+func (a *docTimeStamp) Sign(sig *model.PdfSignature, digest model.Hasher) error {
+	buffer := digest.(*bytes.Buffer)
+	h := a.hashAlgorithm.New()
+	io.Copy(h, buffer)
+	s := h.Sum(nil)
+	r := timestamp.Request{
+		HashAlgorithm:   a.hashAlgorithm,
+		HashedMessage:   s,
+		Certificates:    true,
+		Extensions:      nil,
+		ExtraExtensions: nil,
+	}
+	data, err := r.Marshal()
+	if err != nil {
+		return err
+	}
+
+	resp, err := http.Post(a.timestampServerURL, "application/timestamp-query", bytes.NewBuffer(data))
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("http status code wiats 200 got %d", resp.StatusCode)
```

Comment:
Please correct the error message. Also, might be more efficient to do the status code check before `ioutil.ReadAll(resp.Body)` as the body is not relevant if the request was not successful. Another option would be to leave the status code check where it is now and also print the response body in order to identify the issue easily (maybe print the response body in a `common.Log.Debug` call for `status != 200`).